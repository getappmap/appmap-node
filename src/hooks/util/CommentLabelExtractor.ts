import { ESTree } from "meriyah";
import { SourceMapConsumer } from "source-map-js";

export default class CommentLabelExtractor {
  private lineToComment = new Map<number, ESTree.Comment>();

  constructor(
    public comments: ESTree.Comment[],
    public sourceMap?: SourceMapConsumer,
  ) {
    const singleLineComments = this.comments.filter((c) => c.type === "SingleLine");
    singleLineComments.forEach((c) => {
      const line = this.locate(c.loc);
      if (line) this.lineToComment.set(line, c);
    });
  }

  private locate(loc: ESTree.Node["loc"]): number | undefined {
    if (!loc) return undefined;
    if (this.sourceMap) {
      const mapped = this.sourceMap.originalPositionFor(loc.start);
      if (mapped?.line) return mapped.line;
    } else return loc.start.line;
  }

  labelsFor(funStartLine: number) {
    const labels = new Set<string>();

    // @label or @labels can be specified in a single line comment
    // within continuous single line comments preceding a function:
    //   // @labels l1 l2 l3
    //   // @label l4
    //   function f() { ... }
    for (let line = funStartLine - 1; line > 0; line--) {
      const comment = this.lineToComment.get(line);
      if (!comment) break;

      const match = comment.value.match(/^\s*@labels?\s+(.*)/);
      const lineLabels = match?.[1].split(/\s+/).filter((item) => item.length > 0);

      lineLabels?.forEach((l) => labels.add(l));
    }

    if (labels.size > 0) return Array.from(labels);
  }
}
