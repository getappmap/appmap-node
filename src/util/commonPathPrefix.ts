import { sep } from "node:path";

export default function commonPathPrefix(paths: string[]): string {
  const [first, ...others] = paths;
  if (!first) return "";

  const isEqual = (a: string, b: string) => {
    return process.platform == "win32"
      ? a.localeCompare(b, "en", { sensitivity: "base" }) == 0
      : a == b;
  };

  let prefixLen = 0;
  for (let i = 0; i < first.length; i++)
    if (!others.every((path) => isEqual(path[i], first[i]))) break;
    // We occasionally convert back slash to forward slash even in Windows
    else if (first[i] === sep || first[i] === "/") prefixLen = i;

  return first.slice(0, prefixLen + 1);
}
