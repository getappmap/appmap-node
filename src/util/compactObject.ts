/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
export default function compactObject<T extends object>(x: T): T {
  const result: T = { ...x };
  for (const [k, v] of Object.entries(result))
    if (v === undefined || v === null) delete (result as any)[k];
  return result;
}
