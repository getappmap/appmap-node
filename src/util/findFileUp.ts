import { existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { cwd } from "node:process";

export default function locateFileUp(filename: string, dir = cwd()): string | undefined {
  if (existsSync(join(dir, filename))) return dir;
  const parent = dirname(dir);
  if (parent === dir) return;
  return locateFileUp(filename, parent);
}
