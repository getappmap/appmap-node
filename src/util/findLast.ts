export default function findLast<T, P extends (v: T) => boolean>(
  array: readonly T[],
  predicate: P,
): PredResult<P, T> | undefined {
  let i = array.length;

  while (i > 0) {
    const x = array[--i];
    if (predicate(x)) return x as PredResult<P, T>;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PredResult<P, T> = P extends (v: any) => v is infer R ? R : T;
