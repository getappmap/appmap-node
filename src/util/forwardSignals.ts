import type { ChildProcess } from "child_process";

export default function forwardSignals(
  proc: Pick<ChildProcess, "kill">,
  signals = DEFAULT_FORWARD,
) {
  const forward = proc.kill.bind(proc);
  for (const signal of signals) process.on(signal, forward);
}

const DEFAULT_FORWARD: NodeJS.Signals[] = [
  "SIGABRT",
  "SIGALRM",
  "SIGBUS",
  "SIGCONT",
  "SIGFPE",
  "SIGHUP",
  "SIGILL",
  "SIGINT",
  "SIGIO",
  "SIGIOT",
  "SIGPIPE",
  "SIGPOLL",
  "SIGPROF",
  "SIGPWR",
  "SIGQUIT",
  "SIGSEGV",
  "SIGSTKFLT",
  "SIGSYS",
  "SIGTERM",
  "SIGTRAP",
  "SIGTSTP",
  "SIGTTIN",
  "SIGTTOU",
  "SIGUNUSED",
  "SIGURG",
  "SIGUSR1",
  "SIGUSR2",
  "SIGVTALRM",
  "SIGWINCH",
  "SIGXCPU",
  "SIGXFSZ",
  "SIGBREAK",
  "SIGLOST",
  "SIGINFO",
];
