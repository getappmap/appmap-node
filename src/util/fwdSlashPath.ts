export default function fwdSlashPath(path: string): string {
  if (process.platform != "win32") return path;
  return path?.replaceAll("\\", "/");
}
