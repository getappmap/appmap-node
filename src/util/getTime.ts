import { hrtime } from "node:process";

export function getTime(): number {
  const [sec, nano] = hrtime();
  return sec + nano / 1000000000;
}

export function getTimeInMilliseconds(): number {
  return Number(hrtime.bigint() / 1_000_000n);
}
