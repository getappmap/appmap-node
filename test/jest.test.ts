import { format } from "node:util";

import { integrationTest, readAppmaps, runAppmapNode, runAppmapNodeWithOptions } from "./helpers";

integrationTest("mapping Jest tests", () => {
  expect(runAppmapNode("yarn", "jest", "calc", "--color").status).toBe(1);
  expect(runAppmapNode("yarn", "jest", "asyncLib", "--color").status).toBe(0);
  expect(readAppmaps()).toMatchSnapshot();
});

integrationTest("mapping Jest tests with process recording active", () => {
  expect(
    runAppmapNodeWithOptions(
      {
        env: { ...process.env, APPMAP_RECORDER_PROCESS_ALWAYS: "true" },
      },
      "yarn",
      "jest",
      "calc",
      "--color",
    ).status,
  ).toBe(1);
  const appmaps = readAppmaps();
  expect(appmaps).toMatchSnapshot();

  const appmapsArray = Object.values(appmaps);

  const processMaps = Object.entries(appmaps).filter(
    ([, a]) => a.metadata?.recorder.type == "process",
  );
  if (processMaps.length != 1)
    throw new Error(format("expected one process appmap, got: ", Object.fromEntries(processMaps)));
  expect(appmapsArray.filter((a) => a.metadata?.recorder.type == "tests").length).toEqual(4);
});
