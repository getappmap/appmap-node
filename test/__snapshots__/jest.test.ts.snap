// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping Jest tests 1`] = `
{
  "./tmp/appmap/jest/exception_handling/intentionally_throws.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "test.js:18",
                "name": "errorOut",
                "static": true,
                "type": "function",
              },
            ],
            "name": "test",
            "type": "class",
          },
        ],
        "name": "test",
        "type": "package",
      },
    ],
    "events": [
      {
        "defined_class": "test",
        "event": "call",
        "id": 1,
        "lineno": 18,
        "method_id": "errorOut",
        "parameters": [],
        "path": "test.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "exceptions": [
          {
            "class": "TestError",
            "message": "test error",
            "object_id": 2,
          },
        ],
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "jest-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "exception": {
        "class": "TestError",
        "message": "test error",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "exception handling intentionally throws",
      "recorder": {
        "name": "jest",
        "type": "tests",
      },
      "test_status": "failed",
    },
    "version": "1.12",
  },
  "./tmp/appmap/jest/exception_handling/times_out.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "test.js:22",
                "name": "wait",
                "static": true,
                "type": "function",
              },
            ],
            "name": "test",
            "type": "class",
          },
        ],
        "name": "test",
        "type": "package",
      },
    ],
    "events": [
      {
        "defined_class": "test",
        "event": "call",
        "id": 1,
        "lineno": 22,
        "method_id": "wait",
        "parameters": [
          {
            "class": "Number",
            "name": "ms",
            "value": "20",
          },
        ],
        "path": "test.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 2,
        "parent_id": 1,
        "return_value": {
          "class": "Promise",
          "object_id": 3,
          "value": "Promise { <pending> }",
        },
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "jest-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "exception": {
        "class": "String",
        "message": "Exceeded timeout of 10 ms for a test.
Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout.",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "exception handling times out",
      "recorder": {
        "name": "jest",
        "type": "tests",
      },
      "test_status": "failed",
    },
    "version": "1.12",
  },
  "./tmp/appmap/jest/sub/subtracts_numbers_correctly.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "calc.js:1",
                "name": "sum",
                "static": true,
                "type": "function",
              },
            ],
            "name": "calc",
            "type": "class",
          },
        ],
        "name": "calc",
        "type": "package",
      },
    ],
    "events": [
      {
        "defined_class": "calc",
        "event": "call",
        "id": 1,
        "lineno": 1,
        "method_id": "sum",
        "parameters": [
          {
            "class": "Number",
            "name": "x",
            "value": "1",
          },
          {
            "class": "Number",
            "name": "y",
            "value": "2",
          },
        ],
        "path": "calc.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 2,
        "parent_id": 1,
        "return_value": {
          "class": "Number",
          "value": "3",
        },
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "jest-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "exception": {
        "class": "JestAssertionError",
        "message": "[2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32m-1[39m
Received: [31m3[39m",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "sub subtracts numbers correctly",
      "recorder": {
        "name": "jest",
        "type": "tests",
      },
      "test_status": "failed",
    },
    "version": "1.12",
  },
  "./tmp/appmap/jest/sum/sums_numbers_correctly.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "calc.js:1",
                "name": "sum",
                "static": true,
                "type": "function",
              },
            ],
            "name": "calc",
            "type": "class",
          },
        ],
        "name": "calc",
        "type": "package",
      },
    ],
    "events": [
      {
        "defined_class": "calc",
        "event": "call",
        "id": 1,
        "lineno": 1,
        "method_id": "sum",
        "parameters": [
          {
            "class": "Number",
            "name": "x",
            "value": "1",
          },
          {
            "class": "Number",
            "name": "y",
            "value": "2",
          },
        ],
        "path": "calc.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 2,
        "parent_id": 1,
        "return_value": {
          "class": "Number",
          "value": "3",
        },
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "jest-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "sum sums numbers correctly",
      "recorder": {
        "name": "jest",
        "type": "tests",
      },
      "test_status": "succeeded",
    },
    "version": "1.12",
  },
}
`;
